zipline/data/fx/hdf5.py

```diff
@@ -203,56 +201,55 @@ def get_rates(self, rate, quote, bases, dts):
                 .format(rate, quote)
             )
 
-        # OPTIMIZATION: Row indices correspond to dates, which must be in
+        # OPTIMIZATION: Column indices correspond to dates, which must be in
         # sorted order. Rather than reading the entire dataset from h5, we can
-        # read just the interval from min_row to max_row inclusive
+        # read just the interval from min_col to max_col inclusive
         #
         # However, we also need to handle two important edge cases:
         #
-        #   1. row_ixs contains -1 for dts before the start of self.dts.
-        #   2. col_ixs contains -1 for any currencies we don't know about.
+        #   1. row_ixs contains -1 for any currencies we don't know about.
+        #   2. col_ixs contains -1 for dts before the start of self.dts.
         #
         # If either of the above cases obtains, we want to return NaN for the
         # corresponding output locations.
 
         # We handle (1) by reading raw data into a buffer with one extra
-        # row. When we then apply the row index to permute the raw data into
-        # the correct order, any rows with values of -1 will pull from the
-        # extra row, which will always contain NaN>
+        # column. When we then apply the column index to permute the raw data
+        # into the correct order, any columnss with values of -1 will pull from
+        # the extra column, which will always contain NaN>
         #
-        # We handle (2) by overwriting columns with indices of -1 with NaN as a
+        # We handle (2) by overwriting rows with indices of -1 with NaN as a
         # postprocessing step.
-        slice_begin = max(row_ixs[0], 0)
-        slice_end = max(row_ixs[-1], 0) + 1  # +1 to be inclusive of end date.
+        slice_begin = max(col_ixs[0], 0)
+        slice_end = max(col_ixs[-1], 0) + 1  # +1 to be inclusive of end date.
 
-        # Allocate a buffer full of NaNs with one extra row/column. See
+        # Allocate a buffer full of NaNs with one extra column. See
         # OPTIMIZATION notes above.
         buf = np.full(
-            (slice_end - slice_begin + 1, len(self.currencies)),
+            (len(self.currencies), slice_end - slice_begin + 1),
             np.nan,
         )
 
-        # Read data into all but the last row/column of the buffer.
-        dataset.read_direct(
-            buf[:-1],
-            np.s_[slice_begin:slice_end],
-        )
+        buf[:, :-1] = dataset[:, slice_begin:slice_end]
```

Comment:
If we're not using `read_direct` anymore, we might as well allocate both an extra row and an extra column and use the extra row/column trick for handling both cases rather than using it for just one of them.

Reply:
:+1: